### PR TITLE
Feat: added structure for auto join channel

### LINKED ISSUE #3158: Automatically connect to specific channels
For liveops purposes, we would like to force players to join a channel automatically when they enter Decentraland. For example, if we are celebrating the metaverse festival, we would like everybody to join #metaverse-festival automatically, or by default it would be great to have #help on your first round. 

Of course, people should be able to exit those channels and not be reconnected to them automatically again, so we need to store when a player have decided to quit an automatic channel.

Also, automatic channels can become very spammy. Decentraland operatos will be able to set the notifications on and off for those channels definitions.

**Acceptance Criteria**
- A decentraland operator is able to define a set of channels to be connected automatically (maybe Unleash or similar)
- Together with the name of the channel, it will define if notifications are enabled or not for that channel.
- Players will be connected to the list of automatic channels defined by the operator.
- This channels wont be part of the players "connected channels" list when they start. This means if the channel is no longer in the list of automatic channels, the player wont be reconnected to it in the next session
- Automatic channels wont reduce the number of maximum channels allowed for the player.
- If a player exits an automatic channel, it wont be reconnected to it again on the next session unless the player manually joins the channel.

### PR DESCRIPTION
**Pull request by davidejensen**
## What does this PR change?

Fix #3158 
Adds the possibility of defining a feature flag in unleashed to create a list of channels to force users to auto join with the possibility of enabling/disabling the notifications by default

...

## How to test the changes?

1. Go to: https://play.decentraland.zone/?renderer-branch=feat/auto-channel&kernel-branch=fix/channels-messages
2. Verify that the automatically joined channels are the ones typed here: https://features.decentraland.systems/#/features/variants/explorer-automatic_joined_channels
3. Verify that the notifications for those channels are enabled and disabled based on the FF variant definition
4. More info at #3158 

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 1]
```diff
@@ -135,6 +137,60 @@ public void UpdateChannelMembers(string payload)
             OnUpdateChannelMembers?.Invoke(msg.channelId, msg.members);
         }
 
+    // called by kernel
+    [PublicAPI]
+    public void JoinChannelConfirmation(string payload)
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:142]: can you align the identation?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 2]
```diff
@@ -135,6 +137,60 @@ public void UpdateChannelMembers(string payload)
...
+        else
+            channels[channelId] = channel;
+        
+        if(!autoJoinChannelList.Get().Contains(channelId))
+        {
+            OnChannelJoined?.Invoke(channel);
+            OnChannelUpdated?.Invoke(channel);
+        }
+
+        if(autoJoinChannelList.Get().Contains(channelId))
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:164]: there is no need to check if contains, you could do a remove directly

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 3]
```diff
@@ -135,6 +137,60 @@ public void UpdateChannelMembers(string payload)
...
+
+    // called by kernel
+    [PublicAPI]
+    public void JoinChannelError(string payload)
+    {
+        var msg = JsonUtility.FromJson<JoinChannelErrorPayload>(payload);
+        OnJoinChannelError?.Invoke(msg.channelId, (ChannelErrorCode) msg.errorCode);
+
+        if(autoJoinChannelList.Get().Contains(msg.channelId))
+            autoJoinChannelList.Get().Remove(msg.channelId);
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:181]: there is no need to check if contains, you could do a remove directly

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 4]
```diff
@@ -135,6 +137,60 @@ public void UpdateChannelMembers(string payload)
...
+
+    // called by kernel
+    [PublicAPI]
+    public void LeaveChannelError(string payload)
+    {
+        var msg = JsonUtility.FromJson<JoinChannelErrorPayload>(payload);
+        OnChannelLeaveError?.Invoke(msg.channelId, (ChannelErrorCode) msg.errorCode);
+
+        if(autoJoinChannelList.Get().Contains(msg.channelId))
+            autoJoinChannelList.Get().Remove(msg.channelId);
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:192]: there is no need to check if contains, you could do a remove directly

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 5]
```diff
@@ -224,6 +236,14 @@ public void MuteChannelError(string payload)
             OnMuteChannelError?.Invoke(msg.channelId, (ChannelErrorCode) msg.errorCode);
         }
 
+        public void LeaveChannel(string channelId) 
+        {
+            WebInterface.LeaveChannel(channelId);
+
+            if(autoJoinChannelList.Get().Contains(channelId))
+                autoJoinChannelList.Get().Remove(channelId);
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:244]: there is no need to check if contains, you could do a remove directly

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/Channels/AutomaticJoinChannel.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    [Serializable]
+    public class AutomaticJoinChannelList
+    {
+        public AutomaticJoinChannel[] automaticJoinChannelList;
+    }
+
+    [Serializable]
+    public class AutomaticJoinChannel
+    {
+        public string channelName;
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/Channels/AutomaticJoinChannel.cs:15]: i would rename it to `channelId`

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 7]
```diff
@@ -135,6 +137,60 @@ public void UpdateChannelMembers(string payload)
...
+        var channelInfo = msg.channelInfoPayload[0];
+        var channel = new Channel(channelInfo.channelId, channelInfo.name, channelInfo.unseenMessages, channelInfo.memberCount, channelInfo.joined, channelInfo.muted, channelInfo.description);
+        var channelId = channel.ChannelId;
+        
+        if (channels.ContainsKey(channelId))
+            channels[channelId].CopyFrom(channel);
+        else
+            channels[channelId] = channel;
+        
+        if(!autoJoinChannelList.Get().Contains(channelId))
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:158]: i was thinking a different approach, let me know what do you think.
- `IChatController` add a method: `JoinAutomaticChannel(string channelId)`
- `WorldChatWindowController` method `ConnectToAutoJoinChannels` call `chatControllerJoinAutomaticChannel(channelId)`
- `ChatController.JoinAutomaticChannel(string channelId)` should flag the channels that are expected to be joined as automatic:
```
HashSet<string> requestedAutomaticChannelsToJoin; 
public void JoinAutomaticChannel(string channelId)
{
    requestedAutomaticChannelsToJoin.Add(channelId)
    JoinOrCreateChannel(channelId);
}
```
- when `JoinChannelConfirmation` arises:
```
public void JoinChannelConfirmation(string payload)
{
    ...
    var channel = solveChannelObject;
    if (requestedAutomaticChannelsToJoin.Contains(channel.ChannelId))
    {
        OnAutomaticChannelJoined?.Invoke(channel);
        requestedAutomaticChannelsToJoin.Remove(channel.ChannelId);
    }
    else
        OnChannelJoined?.Invoke(channel);

        OnChannelUpdated?.Invoke(channel);

    ... more stuff
}
```

**Review by lorux0 — CHANGES_REQUESTED** [review]

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WorldChatWindowHUD/WorldChatWindowController.cs [hunk 8]
```diff
@@ -272,6 +273,19 @@ private void HandleChatInitialization()
         if (areJoinedChannelsRequestedByFirstTime) return;
         // we do request joined channels as soon as possible to be able to display messages correctly in the notification panel
         RequestJoinedChannels();
+        ConnectToAutoJoinChannels();
+    }
+
+    private void ConnectToAutoJoinChannels()
+    {
+        AutomaticJoinChannelList joinChannelList = channelsFeatureFlagService.GetAutoJoinChannelsList();
+        for(int i=0; i<joinChannelList.automaticJoinChannelList.Length ;i++)
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/WorldChatWindowHUD/WorldChatWindowController.cs:282]: can you run an IDE to correct formatting?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs (hunk 7 above)
Comment (davidejensen) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs]: We could change it like this but I'd say to do it in a second stage when we will need to rework this functionality to add more things

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs [hunk 9]
```diff
@@ -182,21 +147,25 @@ public void JoinChannelConfirmation(string payload)
...
+            
             if (channels.ContainsKey(channelId))
                 channels[channelId].CopyFrom(channel);
             else
                 channels[channelId] = channel;
+            
+            if(!autoJoinChannelList.Get().Contains(channelId))
+            {
+                OnChannelJoined?.Invoke(channel);
+                OnChannelUpdated?.Invoke(channel);
```
Comment (lorux0) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs:161]: why avoid triggering the update of the channel?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs (hunk 7 above)
Comment (lorux0) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/ChatController/ChatController.cs]: can you leave a TODO comment?

**Review by lorux0 — APPROVED** [review]
good work!

**Pull request merged**